agent/logger/global.go

```diff
@@ -0,0 +1,103 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package logger
+
+import (
+	"sync"
+
+	"github.com/cihub/seelog"
+)
+
+var (
+	loggerMux              = &sync.RWMutex{}
+	globalSeelog           seelog.LoggerInterface
+	globalStructuredLogger *structuredLogger
+)
+
+func init() {
+	loggerMux.Lock()
+	defer loggerMux.Unlock()
+	if globalSeelog == nil {
+		globalSeelog = seelog.Current
+		globalStructuredLogger = newStructuredLogger(globalSeelog, DEFAULT_OUTPUT_FORMAT)
+	}
+}
+
+func setGlobalLogger(li seelog.LoggerInterface, logFormat string) {
+	loggerMux.Lock()
+	defer loggerMux.Unlock()
+	globalSeelog = li
+	globalStructuredLogger = newStructuredLogger(li, logFormat)
+}
+
+func getGlobalStructuredLogger() *structuredLogger {
+	loggerMux.RLock()
```

Comment:
Having to lock and defer unlock on every log message seems like it might not be great for performance. If there's no way around locking then maybe we can use an atomic.Int64 value to indicate that the logger has been "initialized" ?

Another thought I had was maybe using a channel that we close once the logger has been initialized (ie reloadConfig is called), so that all the log messages would hang until that happens. Not sure what implications that might have.

Reply:
The lock is more about the race condition/detection. If not protected by a mutex, race detectors will surely detect this at some point or another. With that being said, I don't think this is a performance problem, I have used [Uber's zap ](https://github.com/uber-go/zap) in the past, which is one of the fastest loggers I've profiled; they use a [RWMutex](https://github.com/uber-go/zap/blob/5dab9368974ab1352e4245f9d33e5bce4c23a034/global.go#L47) too and I've tested logging 10K+ entries per second without issues, this of course depends of the hardware used, but the point is that the read lock is almost negligible given the trade-offs of not having it.

The channel is an interesting idea, but I'm afraid is not much better performance-wise since channels use a mutex [internally](https://github.com/golang/go/blob/4fc9565ffce91c4299903f7c17a275f0786734a1/src/runtime/chan.go#L28) 

Probably the only unnecessary performance hit is the `defer` itself, but since golang 1.14 they claim `defer`s have [_almost_ zero](https://golang.org/doc/go1.14#runtime) performance overhead, still, not really needed here.

Reply:
Just to be clear, I still plan to profile/benchmark this change before merging it to prove it's nothing to be concerned about (or not)

Reply:
what about sync.Once? I think that uses what cam's describing  under the hood.

Reply:
but it has a _much_ cleaner api